src/box/applier.cc

```diff
@@ -369,6 +369,189 @@ applier_connection_init(struct iostream *io, const struct uri *uri,
 	}
 }
 
+static void
+applier_run_ballot_triggers(struct applier *applier, bool success)
+{
+	trigger_run(&applier->on_ballot_update, &success);
+}
+
+/**
+ * Perform a IPROTO_VOTE exchange: send a request and decode the response.
+ * Fallback to this method if master isn't avare of "internal.ballot" event.
+ */
+static void
+applier_get_ballot_from_vote(struct iostream *io, struct xrow_header *row,
+			     struct ibuf *ibuf, struct ballot *ballot)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_vote(row);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type == IPROTO_OK) {
+		xrow_decode_ballot_xc(row, ballot);
+		return;
+	}
+	xrow_decode_error(row);
+	struct diag *diag = diag_get();
+	struct error *e = diag_last_error(diag);
+	/*
+	 * Master isn't aware of IPROTO_VOTE request.
+	 * It's OK - we can proceed without it.
+	 */
+	if (box_error_code(e) == ER_UNKNOWN_REQUEST_TYPE) {
+		diag_clear(diag);
+		return;
+	}
+	error_raise(e);
+}
+
+/**
+ * Perform one round of IPROTO_WATCH("internal.ballot") exchange: send the
+ * request / acknowledge that previous notification is accepted, wait for
+ * response.
+ */
+static void
+applier_get_ballot_from_event(struct iostream *io, struct xrow_header *row,
+			      struct ibuf *ibuf, struct ballot *ballot,
+			      int *decoded_fields)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_watch_key(row, internal_ballot_key, IPROTO_WATCH);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type != IPROTO_EVENT) {
+		xrow_decode_error_xc(row);
+	}
+	struct watch_request watch_request;
+	xrow_decode_watch_xc(row, &watch_request);
+	if (watch_request.data == NULL)
+		tnt_raise(ClientError, ER_NO_SUCH_EVENT, internal_ballot_key);
+	xrow_decode_ballot_event_xc(&watch_request, ballot, decoded_fields);
+}
+
+/** Get applier ballot via the newest of available methods. */
+static void
+applier_get_ballot(struct applier *applier)
+{
+	struct xrow_header row;
+	struct iostream io;
+	struct iostream_ctx io_ctx;
+	struct ibuf ibuf;
+	iostream_ctx_clear(&io_ctx);
+	if (iostream_ctx_create(&io_ctx, IOSTREAM_CLIENT, &applier->uri) != 0)
+		diag_raise();
+	iostream_clear(&io);
+	ibuf_create(&ibuf, &cord()->slabc, 1024);
+	auto guard = make_scoped_guard([&] {
+		if (iostream_is_initialized(&io))
+			iostream_close(&io);
+		iostream_ctx_destroy(&io_ctx);
+		ibuf_destroy(&ibuf);
+	});
+
+	struct greeting greeting;
+
+	applier->addr_len = sizeof(applier->addrstorage);
+	applier_connection_init(&io, &applier->uri, &applier->addr,
+				&applier->addr_len, &io_ctx, &greeting);
+	try {
+		while (true) {
+			int decoded_fields = 0;
+			applier_get_ballot_from_event(&io, &row, &ibuf,
+						      &applier->ballot,
+						      &decoded_fields);
+			/*
+			 * Some events received right after master start
+			 * might be empty. Ignore those.
+			 */
+			if (decoded_fields > 0)
+				applier_run_ballot_triggers(applier, true);
+		}
```

Comment:
> I don't understand. How will this loop ever end if the peer supports events?

Only by cancellation of `applier_ballot_watcher`.
My initial idea was that applier will watch the ballot for the whole lifetime. Maybe we could reuse it for something else later on.
But combined with your comment about sending ballot on each wal write, I think we should unwatch the ballot once the particular applier subscribes:
```diff
diff --git a/src/box/applier.cc b/src/box/applier.cc
index 7ab6f02ff..d96d2815b 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -2184,12 +2184,30 @@ applier_thread_data_create(struct applier *applier,
 	return 0;
 }
 
+/** Interrupt the ballot watcher. */
+static void
+applier_unwatch_ballot(struct applier *applier)
+{
+	if (applier->ballot_watcher != NULL) {
+		assert((applier->ballot_watcher->flags &
+			FIBER_IS_JOINABLE) == 0);
+		/* Non-joinable fibers are automatically recycled. */
+		fiber_cancel(applier->ballot_watcher);
+		applier->ballot_watcher = NULL;
+	}
+}
+
 /**
  * Execute and process SUBSCRIBE request (follow updates from a master).
  */
 static void
 applier_subscribe(struct applier *applier)
 {
+	/*
+	 * Applier doesn't need ballot updates once it subscribes. They were
+	 * needed only during bootstrap or join.
+	 */
+	applier_unwatch_ballot(applier);
 	/* Send SUBSCRIBE request */
 	struct iostream *io = &applier->io;
 	struct ibuf *ibuf = &applier->ibuf;
@@ -2332,19 +2350,6 @@ applier_subscribe(struct applier *applier)
 	unreachable();
 }
 
-/** Interrupt the ballot watcher. */
-static void
-applier_unwatch_ballot(struct applier *applier)
-{
-	if (applier->ballot_watcher != NULL) {
-		assert((applier->ballot_watcher->flags &
-			FIBER_IS_JOINABLE) == 0);
-		/* Non-joinable fibers are automatically recycled. */
-		fiber_cancel(applier->ballot_watcher);
-		applier->ballot_watcher = NULL;
-	}
-}
-
 static inline void
 applier_disconnect(struct applier *applier, enum applier_state state)
 {
```

Reply:
>  Besides, it looks like a busy loop - you do the request-response interaction without any waiting in between, but the events are supposed to be waited on. You need to send WATCH once, get one response, and then wait for next responses until you do UNWATCH. Without sending more requests. Or did I misunderstand the code?

Hm, as far as I understood, each event needs a confirmation, which is just another WATCH request.
Master won't send new events if the replica hasn't confirmed previous one.

So on first loop iteration I subscribe to the event, and on each consequent iteration I confirm previous event receipt and wait for the next event.
This is not a busy loop because master won't send anything on consequent event loop iterations unless the event is really updated.

Reply:
Indeed, thanks for the investigation.